documentation/proposals/Proposal - Enhanced Input Events.md

```diff
@@ -0,0 +1,169 @@
+# Summary
+Proposal for adding important missing functionality to input, as well as other enhancements.
+
+# Contributors
+- ThomasMiz
+
+# Current Status
+- [x] Proposed
+- [x] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decisions
+- Event parameters will be turned into readonly structures to prevent long parameter lists in some cases and allow more parameters (such as deltas) to be added in the future without breaking.
+
+# Proposed API
+The only API changes will be to the events presented by IMouse, IKeyboard and possibly other device interfaces.
+
+## Enums
+
+#### KeyModifiers
+Based on [the modifier keys flag from GLFW](https://www.glfw.org/docs/latest/group__mods.html).
+```cs
+[Flags]
+public enum KeyModifiers
+{
+    Shift = 1 << 1,
+    Control = 1 << 2,
+    Alt = 1 << 3,
+    Super = 1 << 4,
+    CapsLock = 1 << 5,
+    NumLock = 1 << 6
+}
+```
+
+## Structs
+
+#### KeyDownEvent
+```cs
+public readonly struct KeyDownEvent
+{
+    public IKeyboard Keyboard { get; }
+    public Key Key { get; }
+    public int KeyCode { get; }
+    public KeyModifiers Modifiers { get; }
+    public bool IsRepeat { get; }
+    
+    public KeyDownEvent(IKeyboard keyboard, Key key, int keyCode, KeyModifiers modifiers, bool isRepeat)
+}
+```
+
+#### KeyUpEvent
+```cs
+public readonly struct KeyUpEvent
+{
+    public IKeyboard Keyboard { get; }
+    public Key Key { get; }
+    public int KeyCode { get; }
+    public KeyModifiers Modifiers { get; }
+    
+    public KeyUpEvent(IKeyboard keyboard, Key key, int keyCode, KeyModifiers modifiers)
+}
+```
+
+#### KeyCharEvent
+```cs
+public readonly struct KeyCharEvent
+{
+    public IKeyboard Keyboard { get; }
+    public char Character { get; }
+    public int KeyCode { get; }
+    
+    public KeyCharEvent(IKeyboard keyboard, char character, int keyCode)
+}
+```
+
+#### MouseMoveEvent
+```cs
+public readonly struct MouseMoveEvent
+{
+    public IMouse Mouse { get; }
+    public Vector2 Position { get; }
+    public Vector2 Delta { get; }
+    
+    public MouseMoveEvent(IMouse mouse, Vector2 position, Vector2 delta)
+}
+```
+
+#### MouseButtonEvent
+```cs
+public readonly struct MouseButtonEvent
+{
+    public IMouse Mouse { get; }
+    public Vector2 Position { get; }
+    public MouseButton Button { get; }
+    public KeyModifiers Modifiers { get; }
+    
+    public MouseButtonEvent(IMouse mouse, Vector2 position, MouseButton button, KeyModifiers modifiers)
+}
+```
+
+#### MouseScrollEvent
+This would replace the current ScrollWheel struct.

```

Comment:
I doubt it, as we'll need to expose the scroll wheel as a property on `IMouse` as well (i.e. outside of the event)

```suggestion
```